pkg/gencosmosdb/cosmosdb/authorizer.go

```diff
@@ -39,13 +42,96 @@ func NewMasterKeyAuthorizer(masterKey string) (Authorizer, error) {
 }
 
 type tokenAuthorizer struct {
-	token string
+	token       string
+	expiration  time.Time
+	cond        *sync.Cond
+	acquiring   bool
+	lastAttempt time.Time
+	getToken    func() (token string, newExpiration time.Time, err error)
+}
+
+func (a *tokenAuthorizer) Authorize(req *http.Request, resourceType, resourceLink string) error {
+	token, err := a.aquireToken()
+	if err != nil {
+		return err
+	}
+	authorizationHeader := fmt.Sprintf("type=aad&ver=1.0&sig=%s", token)
+	currentTime := time.Now().UTC().Format("Mon, 02 Jan 2006 15:04:05 GMT")
+	req.Header.Set("Authorization", authorizationHeader)
+	req.Header.Set("x-ms-date", currentTime)
+
+	return nil
 }
 
-func (a *tokenAuthorizer) Authorize(req *http.Request, resourceType, resourceLink string) {
-	req.Header.Set("Authorization", url.QueryEscape(a.token))
+func NewTokenAuthorizer(token string, expiration time.Time, getToken func() (token string, newExpiration time.Time, err error)) Authorizer {
+	return &tokenAuthorizer{token: token, expiration: expiration, getToken: getToken, cond: sync.NewCond(&sync.Mutex{})}
 }
 
-func NewTokenAuthorizer(token string) Authorizer {
-	return &tokenAuthorizer{token: token}
+// Get returns the underlying resource.
+// If the resource is fresh, no refresh is performed.
+func (a *tokenAuthorizer) aquireToken() (string, error) {
+	// If the resource is expiring within this time window, update it eagerly.
+	// This allows other goroutines to keep running by using the not-yet-expired
+	// resource value while one goroutine updates the resource.
+	const window = 5 * time.Minute   // To update the token if 5 mins left to expire
+	const backoff = 30 * time.Second // Minimum wait time between attempts
+	now, acquire, expired := time.Now(), false, false
+
+	// acquire exclusive lock
+	a.cond.L.Lock()
+	token := a.token
+
+	for {
+		expired = a.expiration.IsZero() || a.expiration.Before(now)
+		if expired {
+			if !a.acquiring {
+				// This goroutine will acquire the token.
+				a.acquiring, acquire = true, true
+				break
+			}
+			// Getting here means that this goroutine will wait for the updated token
+		} else if a.expiration.Add(-window).Before(now) {
+			// The token is valid but is expiring within the time window(5 mins)
+			if !a.acquiring && a.lastAttempt.Add(backoff).Before(now) {
+				// If another goroutine is not acquiring/renewing the token, and none has attempted
+				// to do so within the last 30 seconds, this goroutine will do it
+				a.acquiring, acquire = true, true
+				break
+			}
+			// This goroutine will use the existing token value while another updates it
+			token = a.token
+			break
+		} else {
+			// The token is not close to expiring, this so using its current value
+			token = a.token
+			break
+		}
+		// If we get here, wait for the new token value to be acquired/updated
+		a.cond.Wait()
+	}
+	a.cond.L.Unlock() // Release the lock so no goroutines are blocked
```

Comment:
There's no point unlocking the condition at this point... it would be just as effective to defer the unlock.